scripts/build_climate_projections.py

```diff
@@ -0,0 +1,480 @@
+CMIP6 projections aggregated into IPCC Global Climate Atlas
+
+Outputs
+-------
+
+- ``cutouts/proj-{cutout}.nc"``: confer :ref:`proj`, a cutout modified to account for future
+climate conditions
+
+Description
+-----------
+
+The rule :mod:`build_climate_projections` creates a cutout file which corresponds to
+a requested year in the future. The parameter projections is calculated combining data
+for cutout.nc which is assumed to be representative data sample for the past climate and
+an ensemble of CMIP6 global climate models to account for the future climate.
+
+A morphing approach is implemented meaning that the projection value x_future is calculated
+data for a representative period present_year in the past x_past using a combination
+of shift dx and stretch a:
+
+x_future = x_past + dx + a(x_past - <x_past>|month),
+
+where
+x_future and x_past have the highest temporal resolution,
+<x_past>|month is the monthly average for a considered month.
+
+The procedure is applied for each month to account for seasonality. Snapshots settings
+are used to define for which months to be used in calculating a climate projection.
+
+Methodology notes
+-----------------
+Originally the morphing approach originates from buildings physics, and is still in active
+use for detailed simulation of heat transfer in buildings. See, for example:
+
+J.B. Dias, G.C. da Graça, P.M.M. Soares (2020) Comparison of methodologies for
+generation of future  building thermal energy simulation. Energy and Buildings,
+vol. 206, 109556, https://doi.org/10.1016/j.enbuild.2019.109556.
+
+J. Pulkkinen, J.-N. Louis (2021) Near- and medium-term hourly morphed mean and extreme future temperature datasets for Jyväskylä, Finland, for building thermal energy demand simulations. Data in Brief, vol. 37, 107209, https://doi.org/10.1016/j.dib.2021.107209.
+"""
+
+
+import datetime as dt
+import os
+
+import atlite
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pypsa
+import xarray as xr
+from _helpers import configure_logging, create_logger
+from shapely.geometry import LineString as Line
+from shapely.geometry import Point
+
+logger = create_logger(__name__)
+
+
+def crop_cmip6(cmip6_xr, cutout_xr):
+    """
+    Crop the global-scale CMIP6 dataset to match the cutout extend.
+
+    Parameters
+    ----------
+    cmip6_xr: xarray
+        CMIP6 climate projections loaded as xarray
+    cutout_xr : xarray of a cutout dataset
+        Cutout dataset loaded as xarray
+    """
+
+    # spartial margin needed to avoid data losses during further interpolation
+    d_pad = 1
+
+    cmip6_region = cmip6_xr.sel(
+        lat=slice(
+            min(cutout_xr.coords["y"].values) - d_pad,
+            max(cutout_xr.coords["y"].values) + d_pad,
+        ),
+        lon=slice(
+            min(cutout_xr.coords["x"].values) - d_pad,
+            max(cutout_xr.coords["x"].values) + d_pad,
+        ),
+    )
```

Comment:
very nice!
d_pad could be automatically get from the difference of the values, but definitely ok for now :D

Reply:
Thank you! A really great idea to align with `atlite` approaches.

As for the particular bounds calculation, I have investigated this idea, but would say that it doesn't feel like a clear approach to use "magic" index numbers. Agree that it's quite common. But personally, I'm not capable to keep in mind which integer means what, and that is usually quite tricky to deal with the code using such conventions. Not sure it's justifiable but performance gains, as currently performance bottlenecks seem to be in spatial calculations. 

What would potentially be interesting to adopt from `atlite` is work with data chunks. It may be probably a good idea to investigate it for increasing the performance. What do you think?

Reply:
Mmm maybe it would be good to at least create here an auxiliary function that, having as input an xarray, returns the boundaries like (xmin, xmax, ymin, ymax) or alike.
having that function may help clarifying the readability regardless of their implementation; although, I believe the xmin may be the first value of x and the last one the maximum.
Likewise, an auxiliary function that calculate the dx and dy may be useful although trivial: the advantage may be readability and manutentibility.

What do you think?